File: src/backpressure.ts
import { ensureWasmLoaded, wasmModule } from './wasm-loader';

export class CreditController {
  private credits: number;
  private readonly initialCredits: number;
  private wasmController?: any;

  constructor(initialCredits: number) {
    this.initialCredits = initialCredits;
    this.credits = initialCredits;
    this.initWasm();
  }

  private async initWasm(): Promise<void> {
    try {
      await ensureWasmLoaded();
      if (wasmModule) {
        this.wasmController = new wasmModule.CreditController(this.initialCredits);
      }
    } catch (error) {
      console.warn('WASM credit controller not available, using JS fallback');
    }
  }

  consumeCredit(amount: number): boolean {
    if (this.wasmController) {
      return this.wasmController.consumeCredit(amount);
    }

    if (this.credits >= amount) {
      this.credits -= amount;
      return true;
    }
    return false;
  }

  addCredits(amount: number): void {
    if (this.wasmController) {
      this.wasmController.addCredits(amount);
    } else {
      this.credits = Math.min(this.credits + amount, Number.MAX_SAFE_INTEGER);
    }
  }

  availableCredits(): number {
    if (this.wasmController) {
      return this.wasmController.availableCredits();
    }
    return this.credits;
  }

  reset(): void {
    this.credits = this.initialCredits;
    if (this.wasmController) {
      this.wasmController = new wasmModule.CreditController(this.initialCredits);
    }
  }

  isExhausted(): boolean {
    return this.availableCredits() === 0;
  }

  hasCredits(): boolean {
    return this.availableCredits() > 0;
  }
}

export class MultiStreamCreditManager {
  private streams: Map<string, CreditController> = new Map();
  private defaultCredits: number;

  constructor(defaultCredits: number) {
    this.defaultCredits = defaultCredits;
  }

  registerStream(streamId: string, initialCredits?: number): void {
    const credits = initialCredits ?? this.defaultCredits;
    this.streams.set(streamId, new CreditController(credits));
  }

  unregisterStream(streamId: string): void {
    this.streams.delete(streamId);
  }

  consume(streamId: string, amount: number): boolean {
    const controller = this.streams.get(streamId);
    if (controller) {
      return controller.consumeCredit(amount);
    }
    return false;
  }

  addCredits(streamId: string, amount: number): void {
    const controller = this.streams.get(streamId);
    if (controller) {
      controller.addCredits(amount);
    }
  }

  availableCredits(streamId: string): number | undefined {
    const controller = this.streams.get(streamId);
    return controller?.availableCredits();
  }

  isStreamExhausted(streamId: string): boolean {
    const controller = this.streams.get(streamId);
    return controller?.isExhausted() ?? true;
  }

  activeStreams(): string[] {
    return Array.from(this.streams.keys());
  }

  totalAvailableCredits(): number {
    let total = 0;
    for (const controller of this.streams.values()) {
      total += controller.availableCredits();
    }
    return total;
  }
}

export class AdaptiveBackpressure {
  private currentRate: number;
  private targetLatencyMs: number;
  private minRate: number;
  private maxRate: number;
  private alpha: number;

  constructor(
    initialRate: number,
    targetLatencyMs: number,
    minRate: number,
    maxRate: number
  ) {
    this.currentRate = initialRate;
    this.targetLatencyMs = targetLatencyMs;
    this.minRate = minRate;
    this.maxRate = maxRate;
    this.alpha = 0.2;
  }

  update(observedLatencyMs: number): void {
    const error = this.targetLatencyMs - observedLatencyMs;
    const adjustment = (this.alpha * error) / this.targetLatencyMs;
    
    const newRate = this.currentRate * (1 + adjustment);
    this.currentRate = Math.max(this.minRate, Math.min(this.maxRate, newRate));
  }

  getRate(): number {
    return this.currentRate;
  }

  getDelayMs(): number {
    if (this.currentRate > 0) {
      return Math.floor(1000 / this.currentRate);
    }
    return Number.MAX_SAFE_INTEGER;
  }

  shouldThrottle(currentThroughput: number): boolean {
    return currentThroughput > this.currentRate;
  }

  setAlpha(alpha: number): void {
    this.alpha = Math.max(0, Math.min(1, alpha));
  }
}

export class WindowedRateLimiter {
  private windowSizeMs: number;
  private maxEvents: number;
  private events: number[] = [];

  constructor(windowSizeMs: number, maxEvents: number) {
    this.windowSizeMs = windowSizeMs;
    this.maxEvents = maxEvents;
  }

  tryAcquire(timestampMs?: number): boolean {
    const now = timestampMs ?? Date.now();
    const cutoff = now - this.windowSizeMs;
    
    this.events = this.events.filter(t => t > cutoff);
    
    if (this.events.length < this.maxEvents) {
      this.events.push(now);
      return true;
    }
    
    return false;
  }

  currentRate(timestampMs?: number): number {
    const now = timestampMs ?? Date.now();
    const cutoff = now - this.windowSizeMs;
    const recentEvents = this.events.filter(t => t > cutoff).length;
    
    return (recentEvents * 1000) / this.windowSizeMs;
  }

  reset(): void {
    this.events = [];
  }

  availableSlots(): number {
    const now = Date.now();
    const cutoff = now - this.windowSizeMs;
    const recentEvents = this.events.filter(t => t > cutoff).length;
    return Math.max(0, this.maxEvents - recentEvents);
  }
}

export interface BackpressureStrategy {
  shouldAccept(metrics: BackpressureMetrics): boolean;
  onAccept(metrics: BackpressureMetrics): void;
  onReject(metrics: BackpressureMetrics): void;
}

export interface BackpressureMetrics {
  queueSize: number;
  processingRate: number;
  inputRate: number;
  latencyMs: number;
  memoryUsage: number;
}

export class DynamicBackpressure implements BackpressureStrategy {
  private maxQueueSize: number;
  private targetLatencyMs: number;
  private adaptive: AdaptiveBackpressure;

  constructor(maxQueueSize: number, targetLatencyMs: number) {
    this.maxQueueSize = maxQueueSize;
    this.targetLatencyMs = targetLatencyMs;
    this.adaptive = new AdaptiveBackpressure(100, targetLatencyMs, 10, 1000);
  }

  shouldAccept(metrics: BackpressureMetrics): boolean {
    if (metrics.queueSize >= this.maxQueueSize) {
      return false;
    }

    if (metrics.latencyMs > this.targetLatencyMs * 2) {
      return false;
    }

    return !this.adaptive.shouldThrottle(metrics.inputRate);
  }

  onAccept(metrics: BackpressureMetrics): void {
    this.adaptive.update(metrics.latencyMs);
  }

  onReject(metrics: BackpressureMetrics): void {
    this.adaptive.update(metrics.latencyMs);
  }
}

File: src/backpressure.rs
use std::sync::Arc;
use std::sync::Mutex;
use std::collections::HashMap;

#[derive(Clone)]
pub struct CreditManager {
    credits: Arc<Mutex<u32>>,
    initial_credits: u32,
}

impl CreditManager {
    pub fn new(initial_credits: u32) -> Self {
        Self {
            credits: Arc::new(Mutex::new(initial_credits)),
            initial_credits,
        }
    }

    pub fn consume(&mut self, amount: u32) -> bool {
        let mut credits = self.credits.lock().unwrap();
        if *credits >= amount {
            *credits -= amount;
            true
        } else {
            false
        }
    }

    pub fn add(&mut self, amount: u32) {
        let mut credits = self.credits.lock().unwrap();
        *credits = (*credits).saturating_add(amount);
    }

    pub fn available(&self) -> u32 {
        *self.credits.lock().unwrap()
    }

    pub fn reset(&mut self) {
        let mut credits = self.credits.lock().unwrap();
        *credits = self.initial_credits;
    }

    pub fn is_exhausted(&self) -> bool {
        *self.credits.lock().unwrap() == 0
    }

    pub fn has_credits(&self) -> bool {
        *self.credits.lock().unwrap() > 0
    }
}

pub struct MultiStreamCreditManager {
    streams: Arc<Mutex<HashMap<String, CreditManager>>>,
    default_credits: u32,
}

impl MultiStreamCreditManager {
    pub fn new(default_credits: u32) -> Self {
        Self {
            streams: Arc::new(Mutex::new(HashMap::new())),
            default_credits,
        }
    }

    pub fn register_stream(&mut self, stream_id: String, initial_credits: Option<u32>) {
        let mut streams = self.streams.lock().unwrap();
        let credits = initial_credits.unwrap_or(self.default_credits);
        streams.insert(stream_id, CreditManager::new(credits));
    }

    pub fn unregister_stream(&mut self, stream_id: &str) {
        let mut streams = self.streams.lock().unwrap();
        streams.remove(stream_id);
    }

    pub fn consume(&mut self, stream_id: &str, amount: u32) -> bool {
        let mut streams = self.streams.lock().unwrap();
        if let Some(manager) = streams.get_mut(stream_id) {
            manager.consume(amount)
        } else {
            false
        }
    }

    pub fn add_credits(&mut self, stream_id: &str, amount: u32) {
        let mut streams = self.streams.lock().unwrap();
        if let Some(manager) = streams.get_mut(stream_id) {
            manager.add(amount);
        }
    }

    pub fn available_credits(&self, stream_id: &str) -> Option<u32> {
        let streams = self.streams.lock().unwrap();
        streams.get(stream_id).map(|m| m.available())
    }

    pub fn is_stream_exhausted(&self, stream_id: &str) -> bool {
        let streams = self.streams.lock().unwrap();
        streams.get(stream_id).map_or(true, |m| m.is_exhausted())
    }

    pub fn active_streams(&self) -> Vec<String> {
        let streams = self.streams.lock().unwrap();
        streams.keys().cloned().collect()
    }

    pub fn total_available_credits(&self) -> u32 {
        let streams = self.streams.lock().unwrap();
        streams.values().map(|m| m.available()).sum()
    }
}

pub struct AdaptiveBackpressure {
    current_rate: Arc<Mutex<f64>>,
    target_latency_ms: f64,
    min_rate: f64,
    max_rate: f64,
    alpha: f64,
}

impl AdaptiveBackpressure {
    pub fn new(
        initial_rate: f64,
        target_latency_ms: f64,
        min_rate: f64,
        max_rate: f64,
    ) -> Self {
        Self {
            current_rate: Arc::new(Mutex::new(initial_rate)),
            target_latency_ms,
            min_rate,
            max_rate,
            alpha: 0.2,
        }
    }

    pub fn update(&mut self, observed_latency_ms: f64) {
        let mut rate = self.current_rate.lock().unwrap();
        
        let error = self.target_latency_ms - observed_latency_ms;
        let adjustment = self.alpha * error / self.target_latency_ms;
        
        let new_rate = *rate * (1.0 + adjustment);
        *rate = new_rate.max(self.min_rate).min(self.max_rate);
    }

    pub fn get_rate(&self) -> f64 {
        *self.current_rate.lock().unwrap()
    }

    pub fn get_delay_ms(&self) -> u64 {
        let rate = self.get_rate();
        if rate > 0.0 {
            (1000.0 / rate) as u64
        } else {
            u64::MAX
        }
    }

    pub fn should_throttle(&self, current_throughput: f64) -> bool {
        current_throughput > self.get_rate()
    }
}

pub struct WindowedRateLimiter {
    window_size_ms: u64,
    max_events: u32,
    events: Arc<Mutex<Vec<u64>>>,
}

impl WindowedRateLimiter {
    pub fn new(window_size_ms: u64, max_events: u32) -> Self {
        Self {
            window_size_ms,
            max_events,
            events: Arc::new(Mutex::new(Vec::new())),
        }
    }

    pub fn try_acquire(&mut self, timestamp_ms: u64) -> bool {
        let mut events = self.events.lock().unwrap();
        
        let cutoff = timestamp_ms.saturating_sub(self.window_size_ms);
        events.retain(|&t| t > cutoff);
        
        if events.len() < self.max_events as usize {
            events.push(timestamp_ms);
            true
        } else {
            false
        }
    }

    pub fn current_rate(&self, timestamp_ms: u64) -> f64 {
        let events = self.events.lock().unwrap();
        let cutoff = timestamp_ms.saturating_sub(self.window_size_ms);
        let recent_events = events.iter().filter(|&&t| t > cutoff).count();
        
        (recent_events as f64 * 1000.0) / self.window_size_ms as f64
    }

    pub fn reset(&mut self) {
        let mut events = self.events.lock().unwrap();
        events.clear();
    }
}